File: Basics/enqueue-dequeue/en-dequeue.js
class Queue {
    constructor() {
        this.items = [];
    }

    enqueue(item) {
        this.items.push(item);
    }

    dequeue() {
        if (!this.isEmpty()) {
            return this.items.shift();
        } else {
            throw new Error("dequeue from an empty queue");
        }
    }

    isEmpty() {
        return this.items.length === 0;
    }
}

File: Basics/enqueue-dequeue/en-dequeue.go
package main

import "fmt"

type Queue struct {
    items []int
}

func (q *Queue) enqueue(item int) {
    q.items = append(q.items, item)
}

func (q *Queue) dequeue() int {
    if !q.isEmpty() {
        item := q.items[0]
        q.items = q.items[1:]
        return item
    } else {
        panic("dequeue from an empty queue")
    }
}

func (q *Queue) isEmpty() bool {
    return len(q.items) == 0
}
